### PR TITLE
RavenDB-10374 hook up to unhandledException and unobservedTaskException

### PR DESCRIPTION
**Pull request by gregolsky**

### CODE REVIEW

File: src/Raven.Server/Utils/UnhandledExceptions.cs [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+namespace Raven.Server.Utils
+{
+    public class UnhandledExceptions
+    {
+        public static void Track(Logger logger)
+        {
+            AppDomain.CurrentDomain.UnhandledException += (sender, eventArgs) =>
+            {
+                var exceptionString = $"UnhandledException: { eventArgs.ExceptionObject.ToString() }.";
+                if (logger.IsInfoEnabled)
```
Comment (ayende) [inline src/Raven.Server/Utils/UnhandledExceptions.cs]: Operations, this is killing us

**Review by ayende — COMMENTED** [review]
What about the actual crash? When we have a finalizer error, for example, do we actually log?
Note that our logger is async, so we might not log it in time.

We need something that would flush a _real_ error and wait for it to be written.

File: src/Raven.Server/Utils/UnhandledExceptions.cs [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+                    logger.Info(exceptionString);
+            };
+
+            TaskScheduler.UnobservedTaskException += (sender, eventArgs) =>
```
Comment (ayende) [inline src/Raven.Server/Utils/UnhandledExceptions.cs]: This is okay as info, but note that we need to do the check _first_

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Utils/UnhandledExceptions.cs (hunk 1 above)
Comment (ppekrol) [reply src/Raven.Server/Utils/UnhandledExceptions.cs]: Done

File: src/Sparrow/Logging/Logger.cs [hunk 3]
```diff
@@ -8,29 +9,46 @@ public class Logger
...
-        public Logger(LoggingSource parent,string source, string logger)
+        public Logger(LoggingSource parent, string source, string logger)
         {
             _parent = parent;
             _source = source;
             _logger = logger;
         }
 
-        public void Info(string msg, Exception ex = null)
+        public Task Info(string msg, Exception ex = null, bool withWait = false)
```
Comment (ayende) [inline src/Sparrow/Logging/Logger.cs]: Better to have a totally separate overload here.

Comment (ayende) [reply src/Sparrow/Logging/Logger.cs]: Call InfoAsync?

**Review by ayende — CHANGES_REQUESTED** [review]

File: src/Raven.Server/Utils/UnhandledExceptions.cs [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+namespace Raven.Server.Utils
+{
+    public class UnhandledExceptions
+    {
+        private static readonly TimeSpan TimeToWaitForLog = TimeSpan.FromSeconds(15);
+
+        public static void Track(Logger logger)
+        {
+            AppDomain.CurrentDomain.UnhandledException += (sender, args) =>
+            {
```
Comment (ayende) [inline src/Raven.Server/Utils/UnhandledExceptions.cs:15]: This is going to kill the process, it ALWAYS interests us.
That said, are we sure that this actually give us a good error _when running as a service_?

File: src/Raven.Server/Utils/UnhandledExceptions.cs [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+            TaskScheduler.UnobservedTaskException += (sender, args) =>
+            {
+                if (logger.IsInfoEnabled == false)
+                    return;
+
+                if (args.Observed)
+                    return;
+
+                var exception = args.Exception.ExtractSingleInnerException();
+                if (logger.IsInfoEnabled)
```
Comment (ayende) [inline src/Raven.Server/Utils/UnhandledExceptions.cs]: already checked

**Comment by ppekrol** [comment]
Applied fixes for comments.

File: src/Raven.Server/Utils/UnhandledExceptions.cs (hunk 4 above)
Comment (ppekrol) [reply src/Raven.Server/Utils/UnhandledExceptions.cs:15]: We are re-checking that, just in case.

**Comment by ayende** [comment]
didn't push them?

**Comment by ppekrol** [comment]
Amended last commit.

**Pull request merged**